### PR TITLE
XFormBuilder for easy XForms

### PR DESCRIPTION
**Pull request by kaapstorm**
XFormBuilder is intended to be a simple way to create an XForm, and add questions to it.

It was created for importing a CDISC ODM document as a CommCare app. Third-party solutions lacked important features like nested groups, and most required their input to come via an Excel spreadsheet or JSON, which did not seem to be a good fit. I also took a look at [app_manager.xform.XForm](https://github.com/dimagi/commcare-hq/blob/xform_builder/corehq/apps/app_manager/xform.py#L539-539) but it seemed aimed at getting data from xfoms already built by Vellum. So I thought XFormBuilder might be a useful tool to add. Usage is as simple as

``` python
xform = XFormBuilder('Built by XFormBuilder')
xform.add_question('personal', 'Personal Questions', data_type='group')
xform.add_question('name', 'What is your name?', group='personal')
xform.add_question('fav_color', 'What is your favorite color?',
                   group='personal',
                   choices={'r': 'Red', 'g': 'Green', 'b': 'Blue'})
xml = xform.tostring(pretty_print=True, encoding='utf-8')
```

(Please excuse the ugly XML formatting in the doctest. I can't see why `pretty_print=True` is so not pretty.)

@millerdev @snopoke cc @orangejenny @czue

### CODE REVIEW

**Comment by kaapstorm** [comment]
(P.S. I think this deserves more than just a doctest. Groups and nested groups don't have tests yet, nor do all the question data types. Unit tests to follow.)

**Comment by millerdev** [comment]
What do you think of changing the API to avoid the problem of passing an unknown group name?
```py
xform = XFormBuilder('Built by XFormBuilder')
personal = xform.add_question('personal', 'Personal Questions', data_type='group')
personal.add_question('name', 'What is your name?')
```

**Comment by kaapstorm** [comment]
Hmmm. Nice idea! Maybe `add_group` returns a Group object? Differentiating Groups and RepeatGroups would also simplify a couple of lines when adding a question to the form body.

**Comment by czue** [comment]
will let others review, but cool stuff!

File: corehq/apps/app_manager/xform_builder.py [hunk 1]
```diff
@@ -0,0 +1,390 @@
...
+            """
+            Return a question node for a normal question
+
+            >>> node = get_input_question_node('text_question')
+            >>> etree.tostring(node)
+            '<input ref="/data/text_question"><label ref="jr:itext(\'text_question-label\')"/></input>'
+
+            """
+            node_ = etree.Element('input', {'ref': self.get_data_ref(name_, groups_)})
+            node_.append(etree.Element('label', {'ref': "jr:itext('{}')".format(self.get_text_id(name_, groups_))}))
```
Comment (dimagimon) [inline corehq/apps/app_manager/xform_builder.py]: E501 line too long (116 > 115 characters)

**Comment by kaapstorm** [comment]
Added class for groups, and made XML a little prettier. More tests to follow.

File: corehq/apps/app_manager/xform_builder.py [hunk 2]
```diff
@@ -0,0 +1,398 @@
...
+        if choices:
+            for choice_name, choice_label in choices.items():
+                self._translation1.append(get_text_node(name, choice_label, group, choice_name))
+
+    def _append_to_model(self, name, data_type, group=None):
+        if data_type in ('string', 'int', 'double', 'date', 'time', 'dateTime'):
+            bind = E.bind({'nodeset': self.get_data_ref(name, group),
+                           'type': 'xsd:' + data_type})
+        else:
+            bind = E.bind({'nodeset': self.get_data_ref(name, group)})
```
Comment (millerdev) [inline corehq/apps/app_manager/xform_builder.py]: You could omit this `else` caluse. Binds with only `nodeset` are ignored.

File: corehq/apps/app_manager/xform_builder.py [hunk 3]
```diff
@@ -0,0 +1,432 @@
...
+... '''      </input>\\n'''
+... '''    </group>\\n'''
+... '''  </h:body>\\n'''
+... '''</h:html>\\n''')
+True
+
+"""
+import uuid
+from lxml import etree
+from lxml.builder import E, ElementMaker
```
Comment (dimagimon) [inline corehq/apps/app_manager/xform_builder.py]: F401 'ElementMaker' imported but unused

**Comment by kaapstorm** [comment]
Added tests. Pending feedback, I'm happy with this now.

**Comment by millerdev** [comment]
Test failure looks real: https://s3.amazonaws.com/archive.travis-ci.org/jobs/90074401/log.txt

File: corehq/apps/app_manager/xform_builder.py [hunk 4]
```diff
@@ -0,0 +1,432 @@
...
+... '''    </input>\\n'''
+... '''    <group ref="/data/personal">\\n'''
+... '''      <label ref="jr:itext(\'personal-label\')"/>\\n'''
+... '''      <input ref="/data/personal/fav_color">\\n'''
+... '''        <label ref="jr:itext(\'personal/fav_color-label\')"/>\\n'''
+... '''      </input>\\n'''
+... '''    </group>\\n'''
+... '''  </h:body>\\n'''
+... '''</h:html>\\n''')
+True
```
Comment (millerdev) [inline corehq/apps/app_manager/xform_builder.py]: What do you think about converting this doctest into a real unit test? It's hard to tell what went wrong when it fails:
```
Expected:
    True
Got:
    False
```
Also, in general there are [some very good reasons not to use doctests for formal testing](http://nedbatchelder.com/blog/200811/things_i_dont_like_about_doctest.html). I think doctests are well suited to verify that example code (documentation) works, but beyond that they become problematic.

Comment (kaapstorm) [inline corehq/apps/app_manager/xform_builder.py]: Good call, thanks @millerdev

File: corehq/apps/app_manager/tests/test_xform_builder.py [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+
+    def replace_xmlns(self, xml, xmlns):
+        return re.sub(r'http://openrosa\.org/formdesigner/[\w-]{36}', xmlns, xml)
+
+    def test_new_question_group(self):
+        """
+        XFormBuilder.new_question should be able to add a group
+        """
+        self.xform.new_question('personal', 'Personal Questions', data_type='group')
+        self.xform.new_question('name', 'What is your name?', group='personal')
```
Comment (millerdev) [inline corehq/apps/app_manager/tests/test_xform_builder.py]: Is there a reason why this doesn't use `personal.new_group`?

File: corehq/apps/app_manager/xform_builder.py [hunk 6]
```diff
@@ -0,0 +1,432 @@
...
+# cf. https://opendatakit.github.io/odk-xform-spec/#data-types
+ODK_TYPES = (
+    'string', 'int', 'boolean', 'decimal', 'date', 'time', 'dateTime', 'select', 'select1', 'geopoint', 'geotrace',
+    'geoshape', 'binary', 'barcode',
+)
+# CommCare question group types
+GROUP_TYPES = ('group', 'repeatGroup')  # TODO: Support 'questionList'
+# The subset of ODK data types that are XSD data types
+# cf. http://www.w3.org/TR/xmlschema-2/#built-in-datatypes
+XSD_TYPES = ('string', 'int', 'boolean', 'decimal', 'date', 'time', 'dateTime')
```
Comment (millerdev) [inline corehq/apps/app_manager/xform_builder.py]: Nitpick: you could eliminate a tiny bit of duplication by defining this before `ODK_TYPES` and then
```py
ODK_TYPES = XSD_TYPES + ('select', 'select1', ...
```

File: corehq/apps/app_manager/tests/test_xform_builder.py (hunk 5 above)
Comment (kaapstorm) [inline corehq/apps/app_manager/tests/test_xform_builder.py]: I test both ways. See two tests down.

File: corehq/apps/app_manager/xform_builder.py (hunk 6 above)
Comment (kaapstorm) [inline corehq/apps/app_manager/xform_builder.py]: :+1:

**Comment by millerdev** [comment]
I'm satisfied. GTM when tests pass (are we still killing intermediate test runs to reduce the travis backlog?)

**Pull request merged**